# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [29]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [3]:
TransaksiDataKlasifikasi = pd.read_csv("D:/Submission/final_transaksi_clustered.csv")
TransaksiDataKlasifikasi.head()

,TransactionAmount,TransactionType,Location,Channel,CustomerAge,CustomerOccupation,TransactionDuration,LoginAttempts,AccountBalance,Cluster
0,14.09,Debit,San Diego,ATM,70,Doctor,81,1,5112.21,3
1,376.24,Debit,Houston,ATM,68,Doctor,141,1,13758.91,3
2,126.29,Debit,Mesa,Online,19,Student,56,1,1122.35,5
3,184.50,Debit,Raleigh,Online,26,Student,25,1,8569.06,5
4,13.45,Credit,Atlanta,Online,26,Student,198,1,7429.40,9


In [4]:
TransaksiDataKlasifikasi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2512 entries, 0 to 2511
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   TransactionAmount    2512 non-null   float64
 1   TransactionType      2512 non-null   object 
 2   Location             2512 non-null   object 
 3   Channel              2512 non-null   object 
 4   CustomerAge          2512 non-null   int64  
 5   CustomerOccupation   2512 non-null   object 
 6   TransactionDuration  2512 non-null   int64  
 7   LoginAttempts        2512 non-null   int64  
 8   AccountBalance       2512 non-null   float64
 9   Cluster              2512 non-null   int64  
dtypes: float64(2), int64(4), object(4)
memory usage: 196.4+ KB


# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [5]:
# Buat instance LabelEncoder
label_encoder = LabelEncoder()
 
# List kolom kategorikal yang perlu di-encode
categorical_columns = ['TransactionType', 'Location', 'Channel', 'CustomerOccupation']
 
# Encode kolom kategorikal
for column in categorical_columns:
    TransaksiDataKlasifikasi[column] = label_encoder.fit_transform(TransaksiDataKlasifikasi[column])
TransaksiDataKlasifikasi.head()

,TransactionAmount,TransactionType,Location,Channel,CustomerAge,CustomerOccupation,TransactionDuration,LoginAttempts,AccountBalance,Cluster
0,14.09,1,36,0,70,0,81,1,5112.21,3
1,376.24,1,15,0,68,0,141,1,13758.91,3
2,126.29,1,23,2,19,3,56,1,1122.35,5
3,184.50,1,33,2,26,3,25,1,8569.06,5
4,13.45,0,1,2,26,3,198,1,7429.40,9


In [9]:
# Buat instance MinMaxScaler
scaler = MinMaxScaler()
 
# Normalisasi semua kolom numerik
numeric_columns = data.select_dtypes(include=['int64', 'float64']).columns
data[numeric_columns] = scaler.fit_transform(data[numeric_columns])
 
# Pisahkan fitur (X) dan target (y)
X = data.drop(columns=['Cluster'])
y = data['Cluster']
 
# Split data menjadi set pelatihan dan set uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
 
# Tampilkan bentuk set pelatihan dan set uji untuk memastikan split
print(f"Training set shape: X_train={X_train.shape}, y_train={y_train.shape}")
print(f"Test set shape: X_test={X_test.shape}, y_test={y_test.shape}")

Training set shape: X_train=(2009, 9), y_train=(2009,)
Test set shape: X_test=(503, 9), y_test=(503,)


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [19]:
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [20]:
knn = KNeighborsClassifier().fit(X_train, y_train)
dt = DecisionTreeClassifier().fit(X_train, y_train)
rf = RandomForestClassifier().fit(X_train, y_train)
svm = SVC().fit(X_train, y_train)
nb = GaussianNB().fit(X_train, y_train)

print("Model training selesai.")

Model training selesai.


- Algoritma K-Nearest Neighbors (KNN) digunakan untuk mengklasifikasikan data baru berdasarkan kedekatannya dengan data yang sudah diberi label dalam dataset pelatihan. KNN sering digunakan karena kemudahannya dalam pemahaman dan implementasi meskipun pada praktiknya, ia dapat menjadi sangat efektif untuk berbagai masalah klasifikasi.
- Algoritma Decision Tree bekerja dengan membagi data menjadi subset yang lebih kecil berdasarkan fitur tertentu hingga mencapai keputusan akhir pada node daun. Proses ini berlangsung secara bertahap dan sistematis, menjadikannya alat yang kuat untuk memahami struktur data serta membuat keputusan berdasarkan fitur yang relevan.
- Algoritma Random Forest menggabungkan beberapa Decision Tree untuk meningkatkan akurasi prediksi dan mengurangi risiko overfitting. Setiap pohon dalam Random Forest dilatih menggunakan subset acak dari data pelatihan dan subset acak dari fitur yang tersedia. Hasil akhir prediksi ditentukan melalui voting (untuk klasifikasi) atau rata-rata (untuk regresi) dari hasil semua pohon dalam model.
- Support Vector Machine (SVM) bertujuan untuk menemukan hyperplane yang memaksimalkan margin, yaitu jarak antara hyperplane dan titik data terdekat dari setiap kelas.
- Naive Bayes bekerja dengan menghitung kemungkinan suatu data termasuk dalam kelas tertentu berdasarkan dua faktor: kemungkinan awal dari setiap kelas (probabilitas prior) dan kemungkinan fitur dalam data jika kelas tersebut benar (probabilitas likelihood). 

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

In [21]:
def evaluate_model(model, X_test, y_test):
    y_pred = model.predict(X_test)
    cm = confusion_matrix(y_test, y_pred)
    tn, fp, fn, tp = cm.ravel()
    results = {
        'Confusion Matrix': cm,
        'True Positive (TP)': tp,
        'False Positive (FP)': fp,
        'False Negative (FN)': fn,
        'True Negative (TN)': tn,
        'Accuracy': accuracy_score(y_test, y_pred),
        'Precision': precision_score(y_test, y_pred),
        'Recall': recall_score(y_test, y_pred),
        'F1-Score': f1_score(y_test, y_pred)
    }
    return results
 
# Mengevaluasi setiap model dan mengumpulkan hasilnya
results = {
    'K-Nearest Neighbors (KNN)': evaluate_model(knn, X_test, y_test),
    'Decision Tree (DT)': evaluate_model(dt, X_test, y_test),
    'Random Forest (RF)': evaluate_model(rf, X_test, y_test),
    'Support Vector Machine (SVM)': evaluate_model(svm, X_test, y_test),
    'Naive Bayes (NB)': evaluate_model(nb, X_test, y_test)
}
 
# Buat DataFrame untuk meringkas hasil
summary_df = pd.DataFrame(columns=['Model', 'Accuracy', 'Precision', 'Recall', 'F1-Score'])
 
# Isi DataFrame dengan hasil
rows = []
for model_name, metrics in results.items():
    rows.append({
        'Model': model_name,
        'Accuracy': metrics['Accuracy'],
        'Precision': metrics['Precision'],
        'Recall': metrics['Recall'],
        'F1-Score': metrics['F1-Score']
    })
 
# Konversi daftar kamus ke DataFrame
summary_df = pd.DataFrame(rows)
 
# Tampilkan DataFrame
print(summary_df)

                          Model  Accuracy  Precision    Recall  F1-Score
0     K-Nearest Neighbors (KNN)  1.000000        1.0  1.000000  1.000000
1            Decision Tree (DT)  0.998012        1.0  0.888889  0.941176
2            Random Forest (RF)  1.000000        1.0  1.000000  1.000000
3  Support Vector Machine (SVM)  1.000000        1.0  1.000000  1.000000
4              Naive Bayes (NB)  1.000000        1.0  1.000000  1.000000


- K-Nearest Neighbors (KNN) menunjukkan akurasi sebesar 100%. Model ini memiliki precision 100%, recall 100%, dan F1-Score 100%. Dengan semua angka yang dihasilkan sempurna maka model menggunakan K-Nearest Neighbors (KNN) mengalami overfitting. Kemungkinan model terlalu hafal dengan data latih sehingga kurang mampu generalisasi data baru.

- Decision Tree memperoleh akurasi sebesar 99,8%. Precision-nya adalah 100%, recall 0,88%, dan F1-Score 94,11%. Model ini memiliki akurasi yangs sangat tinggi. Angka precision yang sempurna menunjukkan bahwa ketika model mengklasifikasikan transaksi seseorang maka prediksi tersebut pasti benar. Recall yang cukup tinggi menunjukkan bahwa model ini sangat baik dalam mengidentifikasi pelanggan. Nilai F1-Score yang tinggi menandakan model ini tetap stabil dan optimal dalam menangkap data positif tanpa terlalu banyak kesalahan.

- Random Forest tampil dengan akurasi tertinggi sebesar 100%. Model ini memiliki precision 100%, recall 100%, dan F1-Score 100%. Dengan semua angka yang dihasilkan sempurna maka model menggunakan Random Forest mengalami overfitting. Kemungkinan model terlalu hafal dengan data latih sehingga kurang mampu generalisasi data baru.

- Support Vector Machine (SVM) memiliki akurasi 100%. Precision-nya mencapai 100%, recall 100%, dan F1-Score 100%. Dengan semua angka yang dihasilkan sempurna maka model menggunakan Support Vector Machine (SVM) mengalami overfitting. Kemungkinan model terlalu hafal dengan data latih sehingga kurang mampu generalisasi data baru.

- Naive Bayes menunjukkan akurasi 100%. Precision-nya adalah 100%, recall 100%, dan F1-Score 100%. Dengan semua angka yang dihasilkan sempurna maka model menggunakan Naive Bayes mengalami overfitting. Kemungkinan model terlalu hafal dengan data latih sehingga kurang mampu generalisasi data baru.

### Kesimpulan:
Dari evaluasi model, terlihat bahwa Decision Tree adalah model dengan performa terbaik, mengungguli model lainnya dalam hal akurasi, precision, recall, dan F1-Score. Keunggulan ini menunjukkan kemampuannya dalam mengidentifikasi transaksi dengan lebih baik dan akurat. Namun apabila menggunakan model menggunakan K-Nearest Neighbors (KNN), Random Forest, Support Vector Machine (SVM), dan Naive Bayes mengalami overfitting.

## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

### Tuning Decision Tree

In [22]:
dt = DecisionTreeClassifier(random_state=42)

# Definisikan parameter yang akan diuji
param_grid = {
    "max_depth": [3, 5, 10, None],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 5],
    "criterion": ["gini", "entropy"]
}

# Grid Search dengan cross-validation (cv=5)
grid_search = GridSearchCV(dt, param_grid, cv=5, scoring="accuracy", n_jobs=-1)
grid_search.fit(X_train, y_train)

# Tampilkan hasil terbaik
print("Best Parameters:", grid_search.best_params_)
print("Best Accuracy:", grid_search.best_score_)

# Gunakan model terbaik
best_dt = grid_search.best_estimator_

Best Parameters: {'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Accuracy: 1.0


## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [28]:
print("Best Parameters:", grid_search.best_params_)
print("Best Accuracy from GridSearchCV:", grid_search.best_score_)

# Evaluasi model terbaik
y_pred = best_dt.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Test Accuracy for Best Model: {accuracy:.4f}")

# Evaluasi metrik lain
print("Classification Report for Best Model:\n", classification_report(y_test, y_pred))

Best Parameters: {'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Accuracy from GridSearchCV: 1.0
Test Accuracy for Best Model: 0.9980
Classification Report for Best Model:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       494
           1       1.00      0.89      0.94         9

    accuracy                           1.00       503
   macro avg       1.00      0.94      0.97       503
weighted avg       1.00      1.00      1.00       503



### Analsisis evaluasi setelah tuning
- K-Nearest Neighbors (KNN) menunjukkan akurasi sebesar 99,8%. Model ini memiliki precision 100%, recall 100% pada kelas 0 dan 89% pada kelas 1, dan F1-Score 100% pada kelas 0 dan 94% pada kelas 1. Model ini memiliki akurasi yang sangat tinggi. Angka precision yang sempurna menunjukkan bahwa ketika model mengklasifikasikan transaksi maka transaksi tersebut pasti benar. Recall yang tinggi menunjukkan bahwa model ini sangat baik dalam mengidentifikasi transaksi. Nilai F1-Score yang tinggi menandakan model ini tetap stabil dan optimal dalam menangkap data positif tanpa terlalu banyak kesalahan.

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.

- Decision Tree tidak menunjukkan peningkatan akurasi, precision, dan F1-Score setelah tuning. Namun model ini sedikit mengalami peningkatan pada recall. Angka precision yang sempurna menunjukkan bahwa ketika model mengklasifikasikan transaksi maka prediksi tersebut pasti benar. Recall yang sangat tinggi menunjukkan bahwa model ini sangat baik dalam mengidentifikasi transaksi. Nilai F1-Score yang tinggi menandakan model ini tetap stabil dan optimal dalam menangkap data positif tanpa terlalu banyak kesalahan. Untuk model K-Nearest Neighbors (KNN), Random Forest, Support Vector Machine (SVM), dan Naive Bayes masih kurang memenuhi karena terjadi overfitting pada penggunaan model ini. 

- Model terbaik untuk data ini adalah Decision Tree. Namun kelemahan model ini yaitu terlalu fleksibel, sensitif terhadap noise, Pohon yang terlalu besar jika tidak dikendalikan maka pohon bisa tumbuh sangat besar serta menjadi sangat rumit dan dapat memperlambat proses prediksi maupun interpretasi, serta perubahan kecil pada data pelatihan bisa menyebabkan perubahan besar dalam struktur pohon sehingga model bisa tidak stabil

- Rekomendasi tindak lanjutan yaitu perlu peningkatan decision tree apabila nantinya model terlalu fleksibel maka gunakan Random Forest untuk meningkatkan stabilitas, untuk meniingkatkan performa dapat menggunakan boosting seperti XGBoost atau LightGBM dan agar model lebih ringan bisa menggunakan BaggingClassifier.